IRGen: fix enum bit packing on big-endian platforms.

mundaym opened the pull request:
This change modifies spare bit masks so that they are arranged in
the byte order of the target platform. It also modifies and
consolidates the code that gathers and scatters bits into enum
values.

All enum-related validation tests are now passing on IBM Z (s390x)
which is a big-endian platform.

mundaym committed:
IRGen: fix enum bit packing on big-endian platforms.

This change modifies spare bit masks so that they are arranged in
the byte order of the target platform. It also modifies and
consolidates the code that gathers and scatters bits into enum
values.

All enum-related validation tests are now passing on IBM Z (s390x)
which is a big-endian platform.

mundaym commented:
@swift-ci Please test.

mundaym commented:
@swift-ci Please test.

mundaym commented:
@swift-ci Please test.

mundaym commented:
@jckarter Sorry for the size of this PR. I might try and split it up a bit further but I'm not sure how much more is worth doing. What do you think? This is the last change needed to get enums working on big-endian systems.

I've put a couple of comments in the review in places where I'm not sure if the changes I have made are correct. Let me know your thoughts.

Are there any places where you think it might make sense to add more tests?

Thanks a lot.

jckarter commented:
My main concerns:

- On Darwin, we can't change the layouts of things that might be public ABI, so I would try to preserve the existing behavior as much as possible. (Eventually we may be able to improve the layout algorithm in ABI-breaking ways, but it would need to be controlled by compiler version, and that's a lot of infrastructure to build just for this.)

- The existing implementation tries hard not to create non-power-of-two integer types in LLVM IR, because from past experience, they cause a lot of pain in backend crashes, miscompiles, or bad code. It looks like this patch, perhaps accidentally, reintroduces nonstandard integer sizes in the generated code.

mundaym commented:
@swift-ci Please test.

mundaym commented:
@jckarter 

> On Darwin, we can't change the layouts of things that might be public ABI, so I would try to preserve the existing behavior as much as possible. (Eventually we may be able to improve the layout algorithm in ABI-breaking ways, but it would need to be controlled by compiler version, and that's a lot of infrastructure to build just for this.)

I don't think there are any changes to the ABI here (well, the little-endian ABI anyway!), I've tried to ensure the layout stays the same. In this latest patch I've refactored multi-payload completeFixedLayout to make it much closer to the original if that helps confidence.

> The existing implementation tries hard not to create non-power-of-two integer types in LLVM IR, because from past experience, they cause a lot of pain in backend crashes, miscompiles, or bad code. It looks like this patch, perhaps accidentally, reintroduces nonstandard integer sizes in the generated code.

Understood. I've modified the code in the EnumPayload emitScatter/Gather passes so that they don't generate new types (that might not be sized to a power-of-2). This has reduced the number of codegen test changes significantly.

Thanks again!

jckarter commented:
Thanks! Yeah, it's a bit clearer how the new implementation corresponds to the old now, I appreciate it. LGTM.
